tests/test_otel.py

```diff
@@ -0,0 +1,524 @@
+            otel_span_attr http.request $request;
+            return 200 "TRACE-ON";
+        }
+
+        location /context-ignore {
+            otel_span_name context_ignore;
+            proxy_pass http://127.0.0.1:18080/204;
+        }
+
+        location /context-extract {
+            otel_trace_context extract;
+            otel_span_name context_extract;
+            proxy_pass http://127.0.0.1:18080/204;
+        }
+
+        location /context-inject {
+            otel_trace_context inject;
+            otel_span_name context_inject;
+            proxy_pass http://127.0.0.1:18080/204;
+        }
+
+        location /context-propagate {
+            otel_trace_context propagate;
+            otel_span_name context_propagate;
+            proxy_pass http://127.0.0.1:18080/204;
+        }
+
+        location /204 {
+            otel_trace off;
+            add_header "X-Otel-Traceparent" $http_traceparent;
+            add_header "X-Otel-Tracestate" $http_tracestate;
+            return 204;
+        }
+    }
+}
+
+"""
+
+(trace_id, span_id) = ("0af7651916cd43dd8448eb211c80319c", "b9c7c989f97918e1")
+context = {
+    "Traceparent": f"00-{trace_id}-{span_id}-01",
+    "Tracestate": "congo=ucfJifl5GOE,rojo=00f067aa0ba902b7",
+}
+
+# Headers from responses
+_headers = {}
+
+
+def decode_id(span, value):
+    if value in ["trace_id", "span_id"]:
+        return binascii.hexlify(getattr(span, value)).decode("utf-8")
+    return value
+
+
+def span_attr(span, attr, atype):
+    for value in (atrb.value for atrb in span.attributes if atrb.key == attr):
+        return getattr(value, atype)
+
+
+def collect_headers(headers, http_ver):
+    if f"http{http_ver}" not in _headers:
+        _headers[f"http{http_ver}"] = []
+    _headers[f"http{http_ver}"].append(headers)
+
+
+def simple_client(scheme, port, path, logger):
+    def do_get(sock, path):
+        http_send = f"GET {path}\n".encode()
+        logger.debug(f"{http_send=}")
+        sock.sendall(http_send)
+        http_recv = sock.recv(1024)
+        logger.debug(f"{http_recv=}")
+        return http_recv.decode("utf-8")
+
+    with socket.create_connection(("127.0.0.1", port)) as sock:
+        if scheme == "https":
+            ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+            ctx.check_hostname = False
+            ctx.verify_mode = ssl.CERT_NONE
+            with ctx.wrap_socket(sock, server_hostname="127.0.0.1") as ssock:
+                recv = do_get(ssock, path)
+        else:
+            recv = do_get(sock, path)
+    return recv
+
+
+@pytest.fixture
+def batches(trace_service, http_ver):
+    return trace_service.spans[http_ver * 3 : http_ver * 3 + 3]
+
+
+@pytest.fixture
+def span(batches, idx):
+    return batches[idx // 10][0].scope_spans[0].spans[idx % 10]
+
+
+@pytest.fixture
+def headers(http_ver, idx):
+    if http_ver:
+        return _headers.get(f"http{http_ver}")[idx]
+
+
+@pytest.fixture(scope="module")
+def _otelcol(pytestconfig, testdir, logger):
```

Comment:
I would refuse such an approach. test_service does not depend on otelcol. Probably better place for it is conftest.py. There is an agreement if the fixture does not return anything its name should start with an underscore.

Reply:
There is a logical dependency in a sense that all the tests which use "trace_service" also need "otelcol" started. Also, "otelcol" is much closer related to "trace_service", than anything in "conftest.py". While there might be other means to implement this, main idea is to encapsulate "otelcol", so that only specifying dependency on "trace_service" is enough.

Not sure how such "underscore agreement" is helpful, but looks like it was [obsoleted](https://docs.astral.sh/ruff/rules/pytest-incorrect-fixture-name-underscore/#removal).

Reply:
Fixed.